flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/operators/sink/ConstraintEnforcer.java

```diff
@@ -188,14 +201,13 @@ public void processElement(StreamRecord<RowData> element) throws Exception {
         for (int i = 0; i < charFieldIndices.length; i++) {
             final int fieldIdx = charFieldIndices[i];
             final int precision = charFieldPrecisions[i];
-            final String stringValue = rowData.getString(fieldIdx).toString();
+            final BinaryStringData stringData = (BinaryStringData) rowData.getString(fieldIdx);
 
-            if (stringValue.length() > precision) {
+            if (stringData.getJavaObject().length() > precision) {
```

Comment:
I think invoking `stringData.getJavaObject()` is not safe if you don't invoke `stringData.ensureMaterialized()` first, looking at the rest of the code in `BinaryStringData`.

Perhaps to get the string length you need either `getSizeInBytes()` or `numChars()`? `BinaryStringDataUtil#substringSQL` uses the latter.

Reply:
yep, `numChars()` seems to be the safe choice, thx!